options/options.go

```diff
@@ -503,80 +515,282 @@ func (opts *ToolOptions) handleUnknownOption(option string, arg flags.SplitArgum
-		}
-		opts.Username = cs.Username
-		opts.Password = cs.Password
-		// Only set Source if it's not the Go driver default; that means a user must
-		// have passed an authsource option or provided a database in the URI path.
-		if _, ok := cs.Options["authsource"]; ok || cs.Database != "" {
+
+		if opts.Username != "" && cs.Username != "" {
+			if opts.Username != cs.Username {
+				return ConflictingArgsErrorFormat("username", opts.Username, cs.Username, "--username")
+			}
+		}
+		if opts.Username != "" && cs.Username == "" {
+			cs.Username = opts.Username
+		}
+		if opts.Username == "" && cs.Username != "" {
+			opts.Username = cs.Username
+		}
+
+		if opts.Password != "" && cs.Password != "" {
+			if opts.Password != cs.Password {
+				return fmt.Errorf("Invalid Options: Cannot specify different password in connection URI and command-line option")
+			}
+		}
+		if opts.Password != "" && cs.Password == "" {
+			cs.Password = opts.Password
+		}
+		if opts.Password == "" && cs.Password != "" {
+			opts.Password = cs.Password
+		}
+
+		if opts.Source != "" && cs.AuthSourceSet {
+			if opts.Source != cs.AuthSource {
+				return ConflictingArgsErrorFormat("authSource", opts.Source, cs.AuthSource, "--authenticationDatabase")
+			}
+		}
+		if opts.Source != "" && !cs.AuthSourceSet {
+			cs.AuthSource = opts.Source
+		}
+		if opts.Source == "" && cs.AuthSourceSet {
 			opts.Source = cs.AuthSource
 		}
-		opts.Auth.Mechanism = cs.AuthMechanism
+
+		if opts.Mechanism != "" && cs.AuthMechanism != "" {
+			if opts.Mechanism != cs.AuthMechanism {
+				return ConflictingArgsErrorFormat("authMechanism", opts.Mechanism, cs.AuthMechanism, "--authenticationMechanism")
+			}
+		}
+		if opts.Mechanism != "" && cs.AuthMechanism == "" {
+			cs.AuthMechanism = opts.Mechanism
+		}
+		if opts.Mechanism == "" && cs.AuthMechanism != "" {
+			opts.Mechanism = cs.AuthMechanism
+		}
+
 	}
+
 	if opts.enabledOptions.Namespace {
-		if opts.Namespace != nil && opts.Namespace.DB != "" {
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--db")
+
+		if opts.DB != "" && cs.Database != "" {
+			if opts.DB != cs.Database {
+				return ConflictingArgsErrorFormat("database", cs.Database, opts.DB, "--db")
+			}
+		}
+		if opts.DB != "" && cs.Database == "" {
+			cs.Database = opts.DB
+		}
+		if opts.DB == "" && cs.Database != "" {
+			opts.DB = cs.Database
 		}
 	}
 
-	opts.Namespace.DB = cs.Database
 	opts.Direct = (cs.Connect == connstring.SingleConnect)
-	opts.ReplicaSetName = cs.ReplicaSet
 
-	if cs.SSL && !BuiltWithSSL {
+	// check replica set name equality
+	if opts.ReplicaSetName != "" && cs.ReplicaSet != "" {
+		if opts.ReplicaSetName != cs.ReplicaSet {
+			return ConflictingArgsErrorFormat("replica set name", cs.ReplicaSet, opts.Host, "--host")
+		}
+
+	}
+	if opts.ReplicaSetName != "" && cs.ReplicaSet == "" {
+		cs.ReplicaSet = opts.ReplicaSetName
+	}
+	if opts.ReplicaSetName == "" && cs.ReplicaSet != "" {
```

Comment:
This could be an `else if`, but I'm not too worried about it. (Ditto the other places this pattern is used).

Reply:
Alternatively, you can also use switches, which have the same benefit while looking slightly less cluttered.

Reply:
Yeah it's very cluttered currently. I'm hoping we can get rid of most of these once we fix the zero value problem.